File: auth/rails/test/services/vibe_nest_auth_test.rb
require "test_helper"

class VibeNestAuthTest < ActiveSupport::TestCase
  def setup
    @auth = VibeNestAuth.new(issuer: "https://vibenest.net/", client_id: "client-reference", client_secret: nil, redirect_uri: "https://reference.example/auth/vibenest/callback")
  end

  test "authorization uses PKCE S256, state, nonce, and pinned callback" do
    flow = @auth.begin_flow("issuer" => "https://vibenest.net/", "authorization_endpoint" => "https://vibenest.net/oauth/authorize")
    query = URI.decode_www_form(URI(flow[:url]).query).to_h
    assert_equal "S256", query["code_challenge_method"]
    assert_equal flow[:pending]["state"], query["state"]
    assert_equal flow[:pending]["nonce"], query["nonce"]
    assert_equal "https://reference.example/auth/vibenest/callback", query["redirect_uri"]
  end

  test "insecure endpoints are rejected" do
    assert_raises(ArgumentError) { VibeNestAuth.new(issuer: "http://invalid/", client_id: "client", client_secret: nil, redirect_uri: "https://reference.example/callback") }
  end
end

File: auth/laravel/tests/VibeNestAuthTest.php
<?php

namespace Tests;

use App\Services\VibeNestAuth;
use PHPUnit\Framework\TestCase;

final class VibeNestAuthTest extends TestCase
{
    public function testAuthorizationUsesPkceStateAndNonce(): void
    {
        $auth = new VibeNestAuth('https://vibenest.net/', 'client-reference', null, 'https://reference.example/auth/vibenest/callback');
        $flow = $auth->begin(['issuer' => 'https://vibenest.net/', 'authorization_endpoint' => 'https://vibenest.net/oauth/authorize']);
        parse_str(parse_url($flow['url'], PHP_URL_QUERY), $query);
        self::assertSame('S256', $query['code_challenge_method']);
        self::assertSame($flow['pending']['state'], $query['state']);
        self::assertSame($flow['pending']['nonce'], $query['nonce']);
        self::assertSame('https://reference.example/auth/vibenest/callback', $query['redirect_uri']);
        self::assertArrayHasKey('verifier', $flow['pending']);
    }

    public function testInsecureIssuerIsRejected(): void
    {
        $this->expectException(\RuntimeException::class);
        new VibeNestAuth('http://vibenest.invalid/', 'client', null, 'https://reference.example/auth/vibenest/callback');
    }
}
